File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reorderList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: None Do not return anything, modify head in-place instead.
        """
        slow,fast=head,head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        prev=None
        current=slow.next
        slow.next = None
        while current:
            new_node = current.next
            current.next=prev
            prev=current
            current=new_node
        first,second =head,prev
        while second:
            t1,t2=first.next,second.next
            first.next = second
            second.next = t1
            first,second = t1,t2

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode s = head;
        ListNode f = head;
        while( f != null && f.next != null ){
            s = s.next;
            f = f.next.next;
        }
        ListNode ms = s.next; // middle start
        s.next = null; // middle end
        ListNode prev = null;
        ListNode curr = ms;
        while(curr != null){
            ListNode t = curr.next;
            curr.next = prev;
            prev = curr;
            curr = t;
        }
        ListNode fi = head;//final
        ListNode se = prev;
        while( se != null ){
            ListNode t1 = fi.next;
            ListNode t2 = se.next;

            fi.next = se;
            se.next = t1;
            fi = t1;
            se = t2;
        }

        

        
    }
}